### PR TITLE
Charmhub retries

### PR DESCRIPTION
**Pull request by SimonRichardson**
The PR attempts to retry charmhub requests when the server returns the
following status codes:

  - http.StatusBadGateway
  - http.StatusGatewayTimeout
  - http.StatusServiceUnavailable
  - http.StatusTooManyRequests

## QA steps

Unit tests pass.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1928182

### CODE REVIEW

File: charmhub/http_test.go [hunk 1]
```diff
@@ -136,6 +139,74 @@ func (s *RESTSuite) TestGetWithFailure(c *gc.C) {
 	c.Assert(err, gc.Not(jc.ErrorIsNil))
 }
 
+func (s *RESTSuite) TestGetWithFailureRetry(c *gc.C) {
+	var called int
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (achilleasa) [inline charmhub/http_test.go:144]: `defer server.Close()`?

File: charmhub/http_test.go [hunk 2]
```diff
@@ -136,6 +139,74 @@ func (s *RESTSuite) TestGetWithFailure(c *gc.C) {
...
+
+	var result interface{}
+	_, err := client.Get(context.TODO(), base, &result)
+	c.Assert(err, gc.Not(jc.ErrorIsNil))
+	c.Assert(called, gc.Equals, 3)
+}
+
+func (s *RESTSuite) TestGetWithFailureWithoutRetry(c *gc.C) {
+	var called int
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (achilleasa) [inline charmhub/http_test.go:166]: `defer server.Close()`?

File: charmhub/http_test.go [hunk 3]
```diff
@@ -136,6 +139,74 @@ func (s *RESTSuite) TestGetWithFailure(c *gc.C) {
...
+
+	var result interface{}
+	_, err := client.Get(context.TODO(), base, &result)
+	c.Assert(err, gc.Not(jc.ErrorIsNil))
+	c.Assert(called, gc.Equals, 1)
+}
+
+func (s *RESTSuite) TestGetWithNoRetry(c *gc.C) {
+	var called int
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (achilleasa) [inline charmhub/http_test.go:188]: `defer server.Close()`?

**Review by achilleasa — APPROVED** [review]
LGTM; please ensure that the httptest servers are cleaned up before landing

File: charmhub/http_test.go (hunk 1 above)
Comment (SimonRichardson) [reply charmhub/http_test.go:144]: good shout

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**